### ISSUE TITLE
HTTP2 Max Streams aware client pooling

### ISSUE THREAD

**Issue by howardjohn** [hyperium/hyper#3623]
**Is your feature request related to a problem? Please describe.**
We are dealing with an issue where pooled http2 connections are hanging due max streams limits. This is problematic for us, as all our streams are very long lived, so they end up ~indefinitely blocked.

We were hoping to get around this by tracking stream counts as part of the pool logic.

However, neither `current` or `max` stream counts are exposed.

Would it be feasible to expose those (at least Max - current can be tracked in our own code)?

Or is there an alternative solution we should look into?

Previous discussion brought up

> No, the client won't start opening new connections when max streams is reached.
> This is because of RFC 9113 Section 9.1:
> 
> Clients SHOULD NOT open more than one HTTP/2 connection to a given host and port pair, where the host is derived from a URI, a selected alternative service [ALT-SVC], or a configured proxy.

**Describe the solution you'd like**
A few options are plausible.

1. Pool can be configured to automatically open up new connections when connections are "full"
2. Hyper exposes max and current stream counts, so a pooling implementation can easily build this logic themselves. This could look something like `fn is_open() { self.current() < self.max() }` 
3. Hyper exposes the current max streams (or, possibly, arbitrary http2 connection settings information. A pooling implementation can track the current streams themselves. I _think_ this could be done reliably external to hyper, but there may be edge cases I am not considering.

**Additional context**
Here is our current pool: https://github.com/istio/ztunnel/blob/a0d539ee6b395a823c31db8f382048dc1947bcb7/src/proxy/pool.rs#L68

Basically we are operating on `http2::SendRequest`s

Some other libraries built on `h2` expose similar: https://github.com/cloudflare/pingora/blob/20fd391f3e78c9349149f11fae94da9e4657478e/pingora-core/src/connectors/http/v2.rs#L239

**Comment by edwardwc** [comment]
Hey there @howardjohn,

If I understand correctly, you'd like to see how many streams are currently open?

**Comment by howardjohn** [comment]
Ideally the current and the configured max streams on the connection - the max is dynamic based on the client/server SETTINGS. 

Without this we end up blocking when we have two many long lived requests to the same host

**Comment by edwardwc** [comment]
@howardjohn have you considered using an atomic counter that a structure increments or decrements when it is created and dropped to track the count?

**Comment by howardjohn** [comment]
> @howardjohn have you considered using an atomic counter that a structure increments or decrements when it is created and dropped to track the count?

Yes, that is our current workaround. However, that only enables us to know how many active streams we have produced as a client. We cannot find the negotiated max streams on the connection because its not exposed by hyper. So today we are guessing no one will set it <100... but there is no guarantee, and it makes us extremely conservative

**Comment by dswij** [comment]
That's a fair ask. I'll get into it as soon as I have time to spare.

**Comment by hackermondev** [comment]
is it possible for http2 max streams to be exposed in hyper? the pull request you mentioned exposes the method in low-level `h2`, would be nice to also have another method in hyper

**Comment by howardjohn** [comment]
> is it possible for http2 max streams to be exposed in hyper? the pull request you mentioned exposes the method in low-level `h2`, would be nice to also have another method in hyper

I pushed https://github.com/hyperium/hyper/pull/4026

**Comment by dentiny** [comment]
Yeah I met the same issue.

Here's my understanding, forgive if I get it wrong.
In the connection pool implementation, for HTTP/2 whenever we get a usable, unexpired connection, it's **always** reinserted into the queue again, without considering the current ongoing stream number and max allowed stream number; so when the max allowed number reached, we could create a new TCP connection instead of reusing existing conn and queuing requests.
https://github.com/hyperium/hyper-util/blob/8ae9e8b1e338a5b7d35155a0dc3708cfd94bcae2/src/client/legacy/pool.rs#L300

To me, I'd appreciate if we could expose a config `http2_max_allowed_streams` which applies to each TCP connection.
I did a draft PR here: https://github.com/dentiny/hyper-util/pull/1

> is it possible for http2 max streams to be exposed in hyper? the pull request you mentioned exposes the method in low-level h2, would be nice to also have another method in hyper

TBH I'm not sure it's same concept, my understand (could be wrong):
- http2_max_concurrent_streams: a HTTP spec param, which is used to notify HTTP server that new requests will queue
- http2_max_allowed_streams: the config I propose to add as a user config, is used inside of hyper-util connection pool

**Comment by dentiny** [comment]
Also notice libcurl supports max connections to one host: https://curl.se/libcurl/c/CURLMOPT_MAX_HOST_CONNECTIONS.html
I'm wondering if this is something we could try for http2 connection pool (i.e., use connection in round-robin style)?

**Comment by seanmonstar** [comment]
Yea, so the legacy client purposefully only opens one connection to the host because of RFC 9113 Section 9.1.

I believe this is the kind of thing that could be done with the newer [composable pools](https://seanmonstar.com/blog/hyper-util-composable-pools/). I'd likely make a pool stack with a tower load balance layer on top of an h2 singleton pool, which then balances over them.

If we need more general layers in tower or hyper-util, that'd be fine.

**Comment by Catwoman08** [comment]
Not directly actionable. https://github.com/hyperium/hyper/issues/3623#issuecomment-5371660809
